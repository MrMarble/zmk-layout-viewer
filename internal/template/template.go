package template

import (
	"os"

	"gopkg.in/yaml.v3"
)

type Template struct {
	Keyboard Keyboard
}

type Keyboard struct {
	Name    string
	Columns []Column
	Keys    []Key
	Mirror  bool
}

type Column struct {
	Keys int
	Step float64
}

type Key struct {
	Column int
	Row    int
	Step   float64
	H      float64
	W      float64
}

func FromFile(name string) (*Template, error) {
	data, err := os.ReadFile(name)
	if err != nil {
		return nil, err
	}
	return Parse(data)
}

func Parse(data []byte) (*Template, error) {
	tpl := Template{}

	err := yaml.Unmarshal(data, &tpl)
	if err != nil {
		return nil, err
	}
	return &tpl, nil
}
