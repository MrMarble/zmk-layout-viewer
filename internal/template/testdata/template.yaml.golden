keyboard:
  columns:
    - keys: 2
    - keys: 2
